corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,287 @@
+                    )
+                    patient_value = matches[0].value
+                    value_map = self._property_map[prop].value_map
+                    case_value = case.get_case_property(prop)
+                    is_equal = value_map.get(patient_value, patient_value) == case_value
+                    yield weight if is_equal else 0
+                else:
+                    yield 0
+
+        return sum(weights())
+
+    def save_match_id(self, case, case_config, patient):
+        """
+        If we are confident of the patient matched to a case, save
+        the patient's ID to the case.
+        """
+        from casexml.apps.case.mock import CaseBlock
+        from corehq.apps.hqcase.utils import submit_case_blocks
+        from corehq.motech.openmrs.repeater_helpers import PERSON_UUID_IDENTIFIER_TYPE_ID
+
+        case_config_ids = case_config['patient_identifiers']
+        case_update = {}
+        id_type_uuid = PERSON_UUID_IDENTIFIER_TYPE_ID
+        if id_type_uuid in case_config_ids:
+            case_property = case_config_ids[id_type_uuid]['case_property']
+            value = patient['uuid']
+            case_update[case_property] = value
+        for identifier in patient['identifiers']:
+            id_type_uuid = identifier['identifierType']['uuid']
+            if id_type_uuid in case_config_ids:
+                case_property = case_config_ids[id_type_uuid]['case_property']
+                value = identifier['identifier']
+                case_update[case_property] = value
+        case_block = CaseBlock(
+            case_id=case.get_id,
+            create=False,
+            update=case_update,
+        )
+        submit_case_blocks([case_block.as_string()], case.domain)
+
+    def find_patients(self, requests, case, case_config):
+        """
+        Matches cases to patients. Returns a list of patients, each
+        with a confidence score >= self.threshold
+        """
+        from corehq.motech.openmrs.logger import logger
+        from corehq.motech.openmrs.repeater_helpers import search_patients
+
+        self.set_property_map(case_config)
+
+        candidates = {}  # key on OpenMRS UUID to filter duplicates
+        for prop in self.searchable_properties:
+            value = case.get_case_property(prop)
+            if value:
+                response_json = search_patients(requests, value)
+                for patient in response_json['results']:
+                    score = self.get_score(patient, case)
+                    if score >= self.threshold:
+                        candidates[patient['uuid']] = PatientScore(patient, score)
+        if not candidates:
+            logger.info(
+                'Unable to match case "%s" (%s): No candidate patients found.',
+                case.name, case.get_id,
+            )
+            return []
+        patients_scores = sorted(candidates.values(), key=lambda cand: cand.score, reverse=True)
```

Comment:
We can avoid sorting if `len(patients_scores) == 1`, so can move this after this `if` block?

Reply:
lol. Yes, absolutely.